File: best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock-mustafasegf.go
package besttimetobuyandsellstock

func maxProfit(prices []int) int {
	low, max := prices[0], 0

	for _, price := range prices {
		if price < low {
			low = price
		}

		if price-low > max {
			max = price - low
		}
	}

	return max
}

File: best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock-mustafasegf.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        prices
            .into_iter()
            .fold((i32::MAX, 0), |(min, profit), price| match price - min {
                num if num < 0 => (price, profit),
                num if num > profit => (min, num),
                _ => (min, profit),
            })
            .1
    }
}
